libraries/Microsoft.Bot.Builder.Shared/AuthenticationHelper.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Security.Claims;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Connector.Authentication;
+
+namespace Microsoft.Bot.Builder
+{
+    /// <summary>
+    /// Authentication helpers.
+    /// </summary>
+    public static class AuthenticationHelper
+    {
+        /// <summary>
+        /// Call context storage to propagate values throught the request.
+        /// </summary>
+        private static AsyncLocal<BotFrameworkAuthenticationContext> asyncLocal = new AsyncLocal<BotFrameworkAuthenticationContext>();
+
+        /// <summary>
+        /// Gets the request authentication context async.
+        /// </summary>
+        /// <param name="authHeader">The authentication header.</param>
+        /// <param name="httpClient">The HTTP client.</param>
+        /// <returns></returns>
+        public static async Task<BotFrameworkAuthenticationContext> GetRequestAuthenticationContextAsync(string authHeader, HttpClient httpClient)
+        {
+            ClaimsIdentity claimsIdentity;
+
+            if (string.IsNullOrEmpty(authHeader))
+            {
+                claimsIdentity = new ClaimsIdentity(new List<Claim>(), "anonymous");
+            }
+
+            bool usingEmulator = EmulatorValidation.IsTokenFromEmulator(authHeader);
+
+            if (usingEmulator)
+            {
+                claimsIdentity = await EmulatorValidation.GetEmulatorTokenIdentityAsync(authHeader, httpClient);
+            }
+            else
+            {
+                claimsIdentity = await ChannelValidation.GetChannelTokenIdentityAsync(authHeader, httpClient);
+            }
+
+            // For requests from channel App Id is in Audience claim of JWT token. For emulator it is in AppId claim. For 
+            // unauthenticated requests we have anonymouse identity provided auth is disabled.
+            var botAppIdClaim = (claimsIdentity.Claims?.SingleOrDefault(claim => claim.Type == AuthenticationConstants.AudienceClaim)
```

Comment:
If you don't require single semantics, I would suggest optimizing this with `FirstOrDefault` instead. `SingleOrDefault` always has an `O(N)` cost where as `FirstOrDefault` has a worst case of `O(N)`, but can stop short as soon it finds one.